connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1645,6 +1646,8 @@ private void startAndStop(Collection<Callable<Void>> callables) {
             startAndStopExecutor.invokeAll(callables);
         } catch (InterruptedException e) {
             // ignore
+        } catch (RejectedExecutionException re) {
+            log.error("startAndStopExecutor already shutdown or full. Not invoking explicit connector/task shutdown");
```

Comment:
This doesn't address the issue described in the ticket. We don't want to handle a `RejectedExecutionException` here, we want to shut down the connectors and tasks running on the herder before shutting down the `startAndStopExecutor`.

Reply:
The reason I did this was that in case `RejectedExecutionException` does get thrown the rest of the logic in `DistributedHerder#halt` method doesn't get executed. Within halt, we stop the members and eventually when we stop the worker, the connectors and tasks that didn't get stopped, have another opportunity to get stopped. But none of that triggers today FWIU. 

Regarding your suggestion about shutting down connectors/tasks _before_ shutting down the `startAndStopExecutor` I can do that as well but that would ideally need to happen from within the `stop` method IMO. To keep changes at a minimal, I did this. Let me know if it makes sense.

Lastly, I increased the `START_AND_STOP_SHUTDOWN_TIMEOUT_MS` to 5 seconds from current 1s which shouldn't harm much IMO but gives a higher window for the connectors/tasks to close cleanly.

Reply:
@vamossagar12 Sorry for the ambiguous JIRA description. I should have made it more explicit what the bug was beyond the title of the ticket.

Yes, you are correct in observing that one negative effect of this bug is that the remaining portions of the halt aren't completed reliably, and that is definitely something that should be improved. That can either be fixed by catching the exception (what you've implemented) or preventing the exception.

If you prevent the exception, you also fix the other negative effect of the bug, which is that connector stop() methods are never called. That is the effect that I described in the title, and the more important aspect that this PR must resolve.

Reply:
Thanks for the clarification @gharris1727 . Hmm I guess one of the ways to prevent the exception to occur is to stop  connectors/tasks from within the `stop` method instead of the current async way via halt. This way, atleast the connector/task stop requests can be handled before attmepting to shut down the startAndStopExecutor. Is there any approach you have in mind?

Reply:
@vamossagar12 the public-facing contract of the `task.shutdown.graceful.timeout.ms` property is that it is used to control how long we wait before abandoning tasks; its use in worker shutdown logic here is not part of that contract and we can use a higher value for the timeout given to `herderExecutor::awaitTermination` without a KIP.

KIP-611 was abandoned (as is noted in the status on the Wiki page).

Reply:
Oh I assumed that `herderExecutor::awaitTermination` has to use the `task.shutdown.graceful.timeout.ms` config and hence my question about the KIP. I have updated the timeout to 20s. Let me know if that's an ok value?

Regarding KIP-611, thanks for the info. I didn't check it's status before asking.

Reply:
@vamossagar12 can you help me understand why you chose 20 seconds? If someone has configured their worker with a `task.graceful.shutdown.timeout.ms` value larger than 20 seconds, wouldn't that actually increase the likelihood of ungraceful shutdown? Have you looked at the herder's tick thread and startup/shutdown logic and found some evidence that 20 seconds is a reasonable timeout?

Reply:
@C0urante , 20s is a very rough estimate on my part. I looked at the other timeouts in this class, the max seemed to be 10s and chose a value which is still higher than that. 

TBH, I don't know what's a good value for this config because IMO there's no one size fits all value here. I see that for each connector/task callable, the graceful shutdown is set to 5s and the `startAndStopExecutor` has a pool size of 8. It could depend upon factors like how many connectors/tasks are running on the worker for example. And this is when the control enters `halt` method. But let's say for some reason, the connector wants to stop when it just enters the tick method upon startup. At that point, it might need to read through config topic and other things. I am not suggesting that reading the config topic is a time consuming operation but the point I am trying to make is that it's not deterministic and depends upon the cluster where it's running. I can bump it upto a minute but that's still a guesstimate. 
That's why, I still feel we should have the logic to handle `RejectedException` thrown upon stopping along with this high timeout value set. That ways, we could cover all scenarios- the most common ones taken care by the higher timeout AND for the outliers, we still give the worker the opportunity to stop the remaining tasks which couldn't be stopped.

Also, your point about user configuring than `task.graceful.shutdown.timeout.ms` would be ignored by what I did makes sense. I have set the `awaitTermination` timeout to the max between the 2.

Reply:
Thank you for looking into this. It's these kind of details that make maintaining this part of the code base tricky, and I really appreciate the time you've put into diving deep on this one!

I've revisited some of the logic here and it seems like the assumption that the cleanup in `halt` (where we invoke, among other things, `member::stop`) will be allowed to proceed cleanly if we just handle a `RejectedExecutionException` being thrown in `startAndStop` is not quite accurate.

The only way to externally trigger worker shutdown is to begin to terminate the JVM (usually done via `SIGINT`, or ctrl+C). At this point, threads are not given time to gracefully terminate (even if they are non-daemon threads); instead, the only thing that's guaranteed to take place before the JVM exiting is the execution of any shutdown hooks that have been added. In this case, we add a [shutdown hook in the `Connect` class](https://github.com/apache/kafka/blob/67c72596afe58363eceeb32084c5c04637a33831/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Connect.java#L95-L105) that [shuts down the REST server and invokes `Herder::stop`](https://github.com/apache/kafka/blob/67c72596afe58363eceeb32084c5c04637a33831/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Connect.java#L68-L69).

So, even if we catch a `RejectedExecutionException` and handle that gracefully, there's still a good chance that the JVM will terminate before we can successfully invoke (among other things) `Worker::stop`.


With that in mind, here's my proposal:

1. Modify the shutdown logic in `halt` to use `Worker::stopAndAwaitConnectors` and `Worker::stopAndAwaitTasks` (the no-args variants of each), instead of using `startAndStop` at all. This will make it easier to reason about how long graceful shutdown should take, and reduce the risk of a `rejectedExecutionException` being thrown
2. Increase the timeout for `herderExecutor::awaitTermination` to a value large enough to accommodate a read to the end of the config topic (which is currently bounded by `workerSyncTimeoutMs`), plus successfully attempting to stop all connectors and tasks (which should now be bounded by 5 seconds (the hardcoded graceful shutdown timeout for connectors) plus `workerTasksShutdownTimeoutMs`)
3. Add a `catch` clause to the body of `startAndStop` that does swallow a `RejectedExecutionException` and just log a `DEBUG`-level message in that case, but only if we're already in the process of shutdown (i.e., `stopping.get()` is `true`), and throws the exception otherwise

The idea here is to still attempt to gracefully shut down the worker, but also handle the edge case where we time out while trying to do so and the herder tick thread attempts to access a now-closed resource.

Do you think this addresses the issue sufficiently?

Reply:
Thanks @C0urante this is making sense now. I think the important thing here is the fact that the JVM itself can be terminated before getting the chance to do a clean shutdown. So far I was trying to just fix the part where shutdown of connectors/tasks doesn't go through but with what you suggested, it would allow an overall cleaner shutdown. I would make the changes and update the PR.

Reply:
@C0urante i made the changes as discussed. I didn't run the full test suite on my local as I the MM tests take a long time and seem to be failing. Would check the errors thrown here later on.

Reply:
> @C0urante i made the changes as discussed. I didn't run the full test suite on my local as I the MM tests take a long time and seem to be failing. Would check the errors thrown here later on.

looks like connect related changes didn't fail as per the latest run.